README.md
# nacl-blob [![stability][0]][1]
[![npm version][2]][3] [![build status][4]][5] [![windows status][12]][13]
[![downloads][8]][9] [![js-standard-style][10]][11]

Encrypt and decrypt DOM blobs using [nacl-stream](https://github.com/dchest/nacl-stream-js) within service workers.  A port of `nacl-stream-js`'s demo into a consumable module.

- [🌎 Example site](https://nacl-blob.netlify.com)

## Usage

```js
const naclBlob = require('nacl-blob')
const nacl = require('tweetnacl')

const key = nacl.box.keyPair().secretKey
const nonce = nacl.randomBytes(16)

// Any kind of data inside of a blob
const arr = new Uint8Array(10 * 1024 * 1024 + 111)
for (let i = 0; i < arr.length; i++) arr[i] = i & 0xff
const blob = new Blob([arr])

const encryptor = naclBlob.encrypt(key, nonce, blob, (err, encryptedBlob) => {
  if (err) throw (err)
  // some time later
  const decryptor = naclBlob.decrypt(key, nonce, encryptedBlob, (err, decryptedBlob) => {
    if (err) throw (err)
    compareBlobs(blob, decryptedBlob)
  })
  decryptor.on('progress', ({position, length}) => { console.log('decrypting %' + (position / length) * 100) })
})

encryptor.on('progress', ({position, length}) => { console.log('encrypting %' + (position / length) * 100) })


function compareBlobs (a, b) {
  const r1 = new FileReader()
  r1.onload = function () {
    const r2 = new FileReader()
    r2.onload = function () {
      if (r1.result !== r2.result) {
        console.error('blobs differ')
      } else {
        console.log('blobs are equal')
      }
    }
    r2.readAsBinaryString(b)
  }
  r1.readAsBinaryString(a)
}
```

## API

### `encrypt = require('nacl-blob').encrypt`

Import the `encrypt` function.

### `decrypt = require('nacl-blob').decrypt`

Import the `decrypt` function.

### ⚠️ IMPORTANT ⚠️

This module uses a build-time [browserify](http://browserify.org) transform called [`workerify`](https://github.com/shama/workerify).  If you are not using browserify, you can import from the transformed version of the module by importing from the `nacl-blob/dist` path. e.g:

```js
const encrypt = require('nacl-blob/dist').encrypt
const decrypt = require('nacl-blob/dist').decrypt
```

(Wanted, a workerify transform that writes out the contents to a separate bundle)

### `encrypt(key, nonce, blob, [opts], cb)`

Encrypt a [File](file) or [Blob](blob), using a `key` and `nonce`.  Returns an event emitter that can be used to display encryption progress.  The encrypted data will be returned in the callback as a [Blob][blob].

The `key` must be a 32-byte [Uint8Array][uint8] or [Node.js Buffer][nodebuff] (see [github.com/dchest/tweetnacl-js#usage](https://github.com/dchest/tweetnacl-js#usage) for details).

The `nonce` must be a 16-byte [Uint8Array][uint8] or [Node.js Buffer][nodebuff].  

The `blob` must be a [Blob][blob] or [File][file].

Optional `opts` include: 

```js
{
  chunkSize: 1024 * 1024,
  mimeType: blob.type
}
```

The `cb` function will fire when the file/blob has been encrypted and have the the following arguments:

- `err`: Any error that occurred durning encryption.  You should handle this.
- `encryptedBlob`: a [Blob][blob] containing the encrypted data.  This can be securely stored/transmitted along with the `nonce` across insecure networks and decrypted with the `key` (assuming secure key exchange is performed elsewhere).

Returns an [Event Emitter][bus] that you can use to listen for the following events:

- `progress`: An event that emits the the progress of encryption in the following shape:

```js
{
  progress, // in bytes
  length // total bytes
}
```

### `decrypt(key, nonce, encryptedBlob, [opts], [cb])`

Decrypt a [Blob][blob] that was encrypted using a `key` and `nonce`.  Returns an event emitter that can be used to track progress.  The decrypted data will be returned in the callback as a [Blob][blob].

The `key` must be the same 32-byte [Uint8Array][uint8] or [Node.js Buffer][nodebuff] used to encrypt the file. (see [github.com/dchest/tweetnacl-js#usage](https://github.com/dchest/tweetnacl-js#usage))

The `nonce` must be the same 16-byte [Uint8Array][uint8] or [Node.js Buffer][nodebuff] used to encrypt the file.

The `encryptedBlob` must be an encrypted [Blob][blob] or [File][file].

Optional `opts` include:

```js
{
  chunkSize: 1024 * 1024,
  mimeType: encryptedBlob.type
}
```

The `cb` function will fire when the file/blob has been decrypted and have the the following arguments:

- `err`: Any error that occurred durning encryption.  You should handle this.
- `decryptedBlob`: a [Blob][blob] containing the decrypted data.

Returns an [Event Emitter][bus] that you can use to listen for the following events:

- `progress`: An event that emits the the progress of encryption in the following shape:

```js
{
  progress, // in bytes
  length // total bytes
}
```

## See Also

- https://github.com/jedisct1/libsodium.js
- https://github.com/jedisct1/libsodium/issues/475
- https://developer.mozilla.org/en-US/docs/Web/API/Body/blob

## License
[MIT](https://tldrlegal.com/license/mit-license)

[0]: https://img.shields.io/badge/stability-experimental-orange.svg?style=flat-square
[1]: https://nodejs.org/api/documentation.html#documentation_stability_index
[2]: https://img.shields.io/npm/v/nacl-blob.svg?style=flat-square
[3]: https://npmjs.org/package/nacl-blob
[4]: https://img.shields.io/travis/bcomnes/nacl-blob/master.svg?style=flat-square
[5]: https://travis-ci.org/bcomnes/nacl-blob
[8]: http://img.shields.io/npm/dm/nacl-blob.svg?style=flat-square
[9]: https://npmjs.org/package/nacl-blob
[10]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square
[11]: https://github.com/feross/standard
[12]: https://ci.appveyor.com/api/projects/status/roy97n2re724n7rl/branch/master?svg=true
[13]: https://ci.appveyor.com/project/bcomnes/nacl-blob/branch/master
[uint8]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array
[file]: https://developer.mozilla.org/en-US/docs/Web/API/File
[blob]: https://developer.mozilla.org/en-US/docs/Web/API/Blob
[bus]: https://github.com/choojs/nanobus
[nodebuff]: https://nodejs.org/api/buffer.html


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v2.0.3](https://github.com/bcomnes/nacl-blob/compare/v2.0.2...v2.0.3) - 2020-02-10

### Merged

- Update mkdirp to the latest version 🚀 [`#16`](https://github.com/bcomnes/nacl-blob/pull/16)
- Update standard to the latest version 🚀 [`#15`](https://github.com/bcomnes/nacl-blob/pull/15)
- chore(package): update dependency-check to version 4.1.0 [`#13`](https://github.com/bcomnes/nacl-blob/pull/13)
- Update rimraf to the latest version 🚀 [`#14`](https://github.com/bcomnes/nacl-blob/pull/14)
- Update standard to the latest version 🚀 [`#11`](https://github.com/bcomnes/nacl-blob/pull/11)
- Update tape-run to the latest version 🚀 [`#10`](https://github.com/bcomnes/nacl-blob/pull/10)
- Update tape-run to the latest version 🚀 [`#9`](https://github.com/bcomnes/nacl-blob/pull/9)

### Fixed

- chore(package): update dependency-check to version 4.1.0 [`#12`](https://github.com/bcomnes/nacl-blob/issues/12)
- chore(package): update standard to version 12.0.1 [`#5`](https://github.com/bcomnes/nacl-blob/issues/5)

### Commits

- Fix lint [`693b2c3`](https://github.com/bcomnes/nacl-blob/commit/693b2c395f9a17349dabea5bc2e94a12741be06f)
- Update README.md [`e670939`](https://github.com/bcomnes/nacl-blob/commit/e670939dd892e70dc2b97b88dc101561fbf56253)
- chore(package): update tweetnacl to 1.0.3 and rebuild [`a65c585`](https://github.com/bcomnes/nacl-blob/commit/a65c5852184b0c39eaf2b2cfeee498f64c04024a)

## [v2.0.2](https://github.com/bcomnes/nacl-blob/compare/v2.0.1...v2.0.2) - 2018-12-13

### Merged

- chore(package): update nanobus to version 4.4.0 [`#8`](https://github.com/bcomnes/nacl-blob/pull/8)

### Commits

- Add autochangelog and gh-release [`a81b490`](https://github.com/bcomnes/nacl-blob/commit/a81b490d766bb13b219d9f40f75c546039c305ce)

## [v2.0.1](https://github.com/bcomnes/nacl-blob/compare/v2.0.0...v2.0.1) - 2018-10-29

### Merged

- chore(package): update nanobus to version 4.3.5 [`#7`](https://github.com/bcomnes/nacl-blob/pull/7)

### Commits

- Add appveyor tests [`c4d24c8`](https://github.com/bcomnes/nacl-blob/commit/c4d24c8bcb020dc87622a2d83251514c00194dbb)
- Update README.md [`763e8ce`](https://github.com/bcomnes/nacl-blob/commit/763e8cec6db4d33942cb121b670102cc06a54027)
- Update CHANGELOG [`e6ade17`](https://github.com/bcomnes/nacl-blob/commit/e6ade178c4f569e4a8acba7fcdf6eb710499af7e)

## [v2.0.0](https://github.com/bcomnes/nacl-blob/compare/v1.1.2...v2.0.0) - 2018-07-01

### Commits

- Return files instead of blobs, when passed a file [`d9e8efb`](https://github.com/bcomnes/nacl-blob/commit/d9e8efbe1d1e449fd7485811dc6b544a6193ca24)

## [v1.1.2](https://github.com/bcomnes/nacl-blob/compare/v1.1.1...v1.1.2) - 2018-07-01

### Commits

- Delete util.js [`fc5a996`](https://github.com/bcomnes/nacl-blob/commit/fc5a9961f1f38069507723da1ead1a47ca912ffa)
- Changelog [`4e1fea1`](https://github.com/bcomnes/nacl-blob/commit/4e1fea1d0962f50d33b77af87fbd54b9915bd38f)
- Actually export something in dist [`a23fec6`](https://github.com/bcomnes/nacl-blob/commit/a23fec68ca52fbf94e2998898a3023b7d208c4ff)

## [v1.1.1](https://github.com/bcomnes/nacl-blob/compare/v1.1.0...v1.1.1) - 2018-05-27

### Commits

- Decrypt example [`d9c567c`](https://github.com/bcomnes/nacl-blob/commit/d9c567c0015aa1dea44800096a6134eef89ef63a)
- Update README regarding key/nonce generation [`76d7ec7`](https://github.com/bcomnes/nacl-blob/commit/76d7ec7864313cd42505783e4efa0ed98e88dfe0)
- Finish up PoC [`41097d1`](https://github.com/bcomnes/nacl-blob/commit/41097d1f080caf25ec46387baf2dc91760947b44)

## [v1.1.0](https://github.com/bcomnes/nacl-blob/compare/v1.0.0...v1.1.0) - 2018-05-21

### Commits

- Add example [`646f678`](https://github.com/bcomnes/nacl-blob/commit/646f6785c53ae70aa9e99ecbc870589d49460582)
- Add argument checks to the module [`59f77be`](https://github.com/bcomnes/nacl-blob/commit/59f77beedb95d82d876f89f3d1bc5a6a79ab446e)
- Fix website [`380333c`](https://github.com/bcomnes/nacl-blob/commit/380333c2c8674afb2946f2d57039754dce13855a)

## v1.0.0 - 2018-05-20

### Commits

- Try a straight copy [`d7bddf4`](https://github.com/bcomnes/nacl-blob/commit/d7bddf4b8b5ce499c197d213db6167298fa7cefb)
- Prepping for 1.0 [`75ce572`](https://github.com/bcomnes/nacl-blob/commit/75ce572213de3ab0e1997f1d9b8167267aff9d32)
- Checking in basic experiments [`d86c95d`](https://github.com/bcomnes/nacl-blob/commit/d86c95d35dbe7d01f6f7887e407540d8e4e81f4c)


test.js
/* eslint-env browser */
const test = require('tape')
const naclBlob = require('./index')
const nacl = require('tweetnacl')

test('test encryption and decryption', t => {
  const key = nacl.box.keyPair().secretKey
  const nonce = nacl.randomBytes(16)

  const arr = new Uint8Array(10 * 1024 * 1024 + 500)
  for (let i = 0; i < arr.length; i++) arr[i] = i & 0xff
  const blob = new Blob([arr])

  let encryptorProgressRan = false
  let decryptorProgressRan = false

  const encryptor = naclBlob.encrypt(key, nonce, blob, (err, encryptedBlob) => {
    t.error(err, 'encrypt without error')
    const decryptor = naclBlob.decrypt(key, nonce, encryptedBlob, (err, decryptedBlob) => {
      t.error(err, 'decrypt without error')

      t.ok(encryptorProgressRan, 'encryptor event emitter was running')
      t.ok(decryptorProgressRan, 'decryptor event emitter was running')

      compareBlobs(blob, decryptedBlob, (err) => {
        t.error(err, 'original blob equals decrypted blob')
        t.end()
      })
    })
    decryptor.on('progress', ({ position, length }) => {
      decryptorProgressRan = true
      console.log('decrypting %' + Math.floor((position / length) * 100))
    })
  })

  encryptor.on('progress', ({ position, length }) => {
    encryptorProgressRan = true
    console.log('encrypting %' + Math.floor((position / length) * 100))
  })
})

function compareBlobs (a, b, cb) {
  const r1 = new FileReader()
  r1.onload = function () {
    const r2 = new FileReader()
    r2.onload = function () {
      if (r1.result !== r2.result) {
        cb(new Error('blobs are not equal'))
      } else {
        cb(null)
      }
    }
    r2.readAsBinaryString(b)
  }
  r1.readAsBinaryString(a)
}


test/index.js
/* eslint-env browser */
const test = require('tape')
const naclBlob = require('../index')

test('test encryption and decryption', t => {
  const key = new Uint8Array(32)
  const nonce = new Uint8Array(16)
  // var arr = nacl.util.decodeUTF8('Hello, chunky!');
  const arr = new Uint8Array(10 * 1024 * 1024 + 111)
  for (let i = 0; i < arr.length; i++) arr[i] = i & 0xff
  const blob = new Blob([arr])
  const encryptor = naclBlob.encrypt(key, nonce, blob, (err, encryptedBlob) => {
    t.error(err)
    const decryptor = naclBlob.decrypt(key, nonce, encryptedBlob, (err, decryptedBlob) => {
      t.error(err)
      compareBlobs(blob, decryptedBlob)
      t.end()
    })
    decryptor.on('progress', ({position, length}) => { console.log('decrypting %' + (position / length) * 100) })
  })

  encryptor.on('progress', ({position, length}) => { console.log('encrypting %' + (position / length) * 100) })
})

function compareBlobs (a, b) {
  const r1 = new FileReader()
  r1.onload = function () {
    const r2 = new FileReader()
    r2.onload = function () {
      if (r1.result !== r2.result) {
        console.error('blobs differ')
      } else {
        console.log('blobs are equal')
      }
    }
    r2.readAsBinaryString(b)
  }
  r1.readAsBinaryString(a)
}


index.js
/* eslint-env browser  */
const nacl = require('nacl-stream')
const Nanobus = require('nanobus')

function checkArgs (key, nonce, maxChunkLength) {
  if (key.length !== 32) return Error('bad key length, must be 32 bytes')
  if (nonce.length !== 16) return new Error('bad nonce length, must be 16 bytes')
  if (maxChunkLength >= 0xffffffff) return new Error('max chunk length is too large')
  if (maxChunkLength < 16) return new Error('max chunk length is too small')
}

function encrypt (key, nonce, blob, opts, cb) {
  if (!opts) opts = {}
  if (typeof opts === 'function') {
    cb = opts
    opts = {}
  }
  opts = Object.assign({
    chunkSize: 1024 * 1024,
    mimeType: blob.type,
    name: blob.name
  }, opts) // defaults

  const argsNotOk = checkArgs(key, nonce, opts.chunkSize)
  if (argsNotOk) return cb(argsNotOk)

  const bus = new Nanobus()

  const encryptedChunks = []
  let position = 0

  const worker = new Worker('./encrypt-worker.js')

  worker.onmessage = function (ev) {
    switch (ev.data.name) {
      case 'ENCRYPT_START_OK': {
        return postNextChunk()
      }

      case 'ENCRYPT_CHUNK_OK': {
        encryptedChunks.push(ev.data.encryptedChunk)
        bus.emit('progress', { position, length: blob.size })
        if (!ev.data.isLast) {
          return postNextChunk()
        }

        return worker.postMessage({
          name: 'ENCRYPT_FINISH'
        })
      }

      case 'ENCRYPT_FINISH_OK': {
        if (opts.name) {
          return cb(null, new File(encryptedChunks, opts.name, { type: opts.mimeType }))
        } else {
          return cb(null, new Blob(encryptedChunks, { type: opts.mimeType }))
        }
      }
      case 'ENCRYPT_CANCEL_OK': {
        return cb(ev.data.reason)
      }

      default: {
        throw new Error('received unknown message from worker ' + event.data.name)
      }
    }
  }

  function error (reason) {
    worker.postMessage({
      name: 'ENCRYPT_CANCEL',
      reason: reason
    })
  }

  // Reads blob slice contents as Uint8Array, passing it to callback.
  function readBlobSlice (blob, start, end, callback) {
    var reader = new FileReader() // XXX cache reader as the enclosed function's var?
    reader.onerror = function (ev) {
      error(ev) // TODO where is actual error string?
    }
    reader.onload = function () {
      callback(new Uint8Array(reader.result))
    }
    reader.readAsArrayBuffer(blob.slice(start, end))
  }

  // Feeds next chunk to worker and advances position.
  function postNextChunk () {
    var isLast = false
    var end = position + opts.chunkSize
    if (end >= blob.size) {
      end = blob.size
      isLast = true
    }
    readBlobSlice(blob, position, end, function (chunk) {
      worker.postMessage({
        name: 'ENCRYPT_CHUNK',
        chunk: chunk,
        isLast: isLast
      })
      // Advance position.
      position = end
    })
  }

  // Start encryption!
  worker.postMessage({
    name: 'ENCRYPT_START',
    key: key,
    nonce: nonce,
    maxChunkLength: opts.chunkSize
  })

  return bus
}

exports.encrypt = encrypt

function decrypt (key, nonce, blob, opts, cb) {
  if (!opts) opts = {}
  if (typeof opts === 'function') {
    cb = opts
    opts = {}
  }
  opts = Object.assign({
    chunkSize: 1024 * 1024,
    mimeType: blob.type,
    name: blob.name
  }, opts) // defaults

  const argsNotOk = checkArgs(key, nonce, opts.chunkSize)
  if (argsNotOk) return cb(argsNotOk)

  const bus = new Nanobus()

  const decryptedChunks = []
  let position = 0
  let nextChunkSize = -1

  const worker = new Worker('./decrypt-worker.js')

  worker.onmessage = function (ev) {
    switch (ev.data.name) {
      case 'DECRYPT_START_OK': {
        return postNextChunk()
      }

      case 'DECRYPT_CHUNK_OK': {
        if (!ev.data.decryptedChunk) {
          return error('decryption failed')
        }

        decryptedChunks.push(ev.data.decryptedChunk)

        bus.emit('progress', { position, length: blob.size })

        if (!ev.data.isLast) {
          return postNextChunk()
        }

        return worker.postMessage({
          name: 'DECRYPT_FINISH'
        })
      }

      case 'DECRYPT_FINISH_OK': {
        if (opts.name) {
          return cb(null, new File(decryptedChunks, opts.name, { type: opts.mimeType }))
        } else {
          return cb(null, new Blob(decryptedChunks, { type: opts.mimeType }))
        }
      }

      case 'DECRYPT_CANCEL_OK': {
        return cb(ev.data.reason)
      }

      default: {
        throw new Error('received unknown message from worker ' + ev.data.name)
      }
    }
  }

  // Cancel decryption with error.
  function error (reason) {
    worker.postMessage({
      name: 'DECRYPT_CANCEL',
      reason: reason
    })
  }

  // Reads blob slice contents as Uint8Array, passing it to callback.
  function readBlobSlice (blob, start, end, callback) {
    var reader = new FileReader() // XXX cache reader as the enclosed function's var?
    reader.onerror = function (event) {
      error(event) // XXX what's the actual error description?
    }
    reader.onload = function () {
      callback(new Uint8Array(reader.result))
    }
    reader.readAsArrayBuffer(blob.slice(start, end))
  }

  // Feeds next chunk to worker and advances position.
  function postNextChunk () {
    if (nextChunkSize === -1) {
      // We are just starting, so read first chunk length.
      if (position + 2 >= blob.size) {
        return error('blob is too short')
      }
      readBlobSlice(blob, position, position + 4, function (data) {
        nextChunkSize = nacl.stream.readChunkLength(data)
        position = 4
        // Now that we have chunk size, call ourselves again.
        postNextChunk()
      })
    } else {
      // Read next chunk + length of the following chunk after it.
      var isLast = false
      var end = position + nextChunkSize + 16 /* tag */ + 4 /* length */
      if (end >= blob.size) {
        end = blob.size
        isLast = true
      }
      readBlobSlice(blob, position - 4 /* include chunk length */, end, function (chunk) {
        if (!isLast) {
          // Read next chunk's length.
          nextChunkSize = nacl.stream.readChunkLength(chunk, chunk.length - 4)
          // Slice the length off.
          chunk = chunk.subarray(0, chunk.length - 4)
        } else {
          nextChunkSize = 0
        }
        // Decrypt.
        worker.postMessage({
          name: 'DECRYPT_CHUNK',
          chunk: chunk,
          isLast: isLast
        })
        // Advance position.
        position = end
      })
    }
  }

  // Start decryption!
  worker.postMessage({
    name: 'DECRYPT_START',
    key: key,
    nonce: nonce,
    maxChunkLength: opts.chunkSize
  })

  return bus
}

exports.decrypt = decrypt


decrypt-worker.js
/* eslint-env worker */

const nacl = require('nacl-stream')

onmessage = function (ev) {
  switch (ev.data.name) {
    case 'DECRYPT_START': {
      return startDecryption(ev.data.key, ev.data.nonce, ev.data.maxChunkLength)
    }

    case 'DECRYPT_CHUNK': {
      return decryptChunk(ev.data.chunk, ev.data.isLast)
    }

    case 'DECRYPT_FINISH': {
      return finishDecryption()
    }
    case 'DECRYPT_CANCEL': {
      return cancelDecryption(ev.data.reason)
    }

    default: {
      throw new Error('worker received unknown message ' + ev.data.name)
    }
  }
}

let decryptor = null

function startDecryption (key, nonce, maxChunkLength) {
  decryptor = nacl.stream.createDecryptor(key, nonce, maxChunkLength)
  postMessage({
    name: 'DECRYPT_START_OK'
  })
}

function decryptChunk (chunk, isLast) {
  var decryptedChunk = decryptor.decryptChunk(chunk, isLast)
  postMessage({
    name: 'DECRYPT_CHUNK_OK',
    decryptedChunk: decryptedChunk,
    isLast: isLast
  })
}

function finishDecryption () {
  decryptor.clean()
  decryptor = null
  postMessage({
    name: 'DECRYPT_FINISH_OK'
  })
}

function cancelDecryption (reason) {
  decryptor.clean()
  decryptor = null
  postMessage({
    name: 'DECRYPT_CANCEL_OK',
    reason: reason
  })
}


encrypt-worker.js
/* eslint-env worker */

const nacl = require('nacl-stream')

onmessage = function (ev) {
  switch (ev.data.name) {
    case 'ENCRYPT_START': {
      return startEncryption(ev.data.key, ev.data.nonce, ev.data.maxChunkLength)
    }

    case 'ENCRYPT_CHUNK': {
      return encryptChunk(ev.data.chunk, ev.data.isLast)
    }

    case 'ENCRYPT_FINISH': {
      return finishEncryption()
    }

    case 'ENCRYPT_CANCEL': {
      return cancelEncryption(ev.data.reason)
    }

    default: {
      throw new Error('worker received unknown message ' + ev.data.name)
    }
  }
}

let encryptor = null

function startEncryption (key, nonce, maxChunkLength) {
  encryptor = nacl.stream.createEncryptor(key, nonce, maxChunkLength)
  postMessage({
    name: 'ENCRYPT_START_OK'
  })
}

function encryptChunk (chunk, isLast) {
  var encryptedChunk = encryptor.encryptChunk(chunk, isLast)
  postMessage({
    name: 'ENCRYPT_CHUNK_OK',
    encryptedChunk: encryptedChunk,
    isLast: isLast
  })
}

function finishEncryption () {
  encryptor.clean()
  encryptor = null
  postMessage({
    name: 'ENCRYPT_FINISH_OK'
  })
}

function cancelEncryption (reason) {
  encryptor.clean()
  encryptor = null
  postMessage({
    name: 'ENCRYPT_CANCEL_OK',
    reason: reason
  })
}


example-site/decrypt-example.js
const { decrypt } = require('../index.js')
const qs = document.querySelector.bind(document)
const nacl = require('tweetnacl')
nacl.util = require('tweetnacl-util')
const fetch = require('node-fetch')

const testDataName = '/encrypted-test-data.gif'

const decryptSelect = qs('#decryptselect')
const decryptFile = qs('#load-decrypt-file')
const decryptKey = qs('#decrypt-key')
const decryptNonce = qs('#decrypt-nonce')
const decryptButton = qs('#decrypt-button')
const decryptMessage = qs('#decrypt-messages')
const decryptImage = qs('#decrypted-image')
const decryptDownload = qs('#decrypted-download-link')

const gifKey = decryptKey.value
const gifNonce = decryptNonce.value

let decryptMode = 'gif'
let uploadFileObj = null
let decryptedPreviewURLObj = null

decryptSelect.addEventListener('change', ev => {
  const source = ev.target.value

  switch (source) {
    case 'gif': {
      decryptMode = 'gif'
      decryptFile.setAttribute('disabled', '')
      decryptKey.setAttribute('disabled', '')
      decryptKey.value = gifKey
      decryptNonce.setAttribute('disabled', '')
      decryptNonce.value = gifNonce
      decryptMessage.innerText = ''
      decryptButton.innerText = 'Download and decrypt'
      decryptButton.removeAttribute('disabled')
      break
    }
    case 'upload':
    default: {
      decryptMode = 'upload'
      decryptFile.removeAttribute('disabled')
      decryptKey.removeAttribute('disabled')
      decryptNonce.removeAttribute('disabled')
      decryptButton.innerText = 'Load and decrypt'
      if (decryptFile.files.length > 0) {
        decryptButton.removeAttribute('disabled')
      } else {
        decryptButton.setAttribute('disabled', '')
      }
      break
    }
  }
})

decryptFile.addEventListener('change', ev => {
  uploadFileObj = ev.currentTarget.files[0]
  if (uploadFileObj) {
    decryptButton.removeAttribute('disabled')
    decryptMessage.innerText = `type: ${uploadFileObj.type} size: ${uploadFileObj.size}`
  }
})

decryptButton.addEventListener('click', ev => {
  decryptButton.setAttribute('disabled', '')
  decryptButton.innerText = 'Loading...'
  if (decryptedPreviewURLObj) URL.revokeObjectURL(decryptedPreviewURLObj)
  switch (decryptMode) {
    case 'gif': {
      fetch(testDataName).then(function (response) {
        if (response.ok) {
          return response.blob()
        }
        throw new Error('Network response was not ok.')
      }).then(encryptedBlob => {
        decryptMessage.innerText = decryptMessage.innerText = 'Got blob'
        return new Promise((resolve, reject) => {
          const key = nacl.util.decodeBase64(decryptKey.value)
          const nonce = nacl.util.decodeBase64(decryptNonce.value)
          const decryptor = decrypt(key, nonce, encryptedBlob, (err, decryptedBlob) => {
            if (err) return reject(err)
            resolve(decryptedBlob)
          })
          decryptor.on('progress', ({ position, length }) => {
            decryptMessage.innerText = `decrypting: ${(position / length) * 100}%`
          })
        })
      }).then(decyptedBlob => {
        decryptedPreviewURLObj = URL.createObjectURL(decyptedBlob)

        if (decyptedBlob.type.startsWith('image')) {
          decryptImage.src = decryptedPreviewURLObj
        }
        decryptDownload.href = decryptedPreviewURLObj
      }).catch(err => {
        decryptMessage.innerText = err.message
      }).then(() => {
        decryptButton.innerText = 'Download and decrypt'
        decryptButton.removeAttribute('disabled')
      })
      break
    }
    case 'upload':
    default: {
      break
    }
  }
})


example-site/encrypt-example.js
/* eslint-env browser */
const { encrypt } = require('../index.js')
const qs = document.querySelector.bind(document)
const nacl = require('tweetnacl')
nacl.util = require('tweetnacl-util')

const fileLoader = qs('#load-file')
const encryptButton = qs('#encrypt-button')
const encryptionMessages = qs('#encrypt-messages')
const inputImage = qs('#input-image')
const encryptionKeyInput = qs('#encryption-key')
const nonceField = qs('#nonce')
const keyUsed = qs('#key-used')
const downloadEncrypted = qs('#download-encrypted')
const encryptionErrors = qs('#encrypt-errors')
const generateEncryptionKeyButton = qs('#generate-encryption-key')

let fileSelected = null

let encryptionKey = nacl.box.keyPair().secretKey
encryptionKeyInput.value = nacl.util.encodeBase64(encryptionKey)
let encrypted = null
let encryptedObjURL = null

generateEncryptionKeyButton.addEventListener('click', () => {
  encryptionKey = nacl.box.keyPair().secretKey
  encryptionKeyInput.value = nacl.util.encodeBase64(encryptionKey)
})

fileLoader.addEventListener('change', ev => {
  fileSelected = ev.currentTarget.files[0]
  if (fileSelected) {
    encryptButton.removeAttribute('disabled')
    encryptionMessages.innerText = `type: ${fileSelected.type} size: ${fileSelected.size}`
    if (fileSelected.type.startsWith('image')) {
      const reader = new FileReader()
      reader.addEventListener('load', function () {
        inputImage.src = reader.result
      })

      reader.readAsDataURL(fileSelected)
    }
  }
})

encryptionKeyInput.addEventListener('input', ev => {
  encryptionKey = nacl.util.decodeBase64(ev.currentTarget.value)
  console.log(encryptionKey)
})

encryptButton.addEventListener('click', ev => {
  if (fileSelected) {
    if (encryptedObjURL) URL.revokeObjectURL(encryptedObjURL)
    const nonce = nacl.randomBytes(16)
    const key = encryptionKey
    const encryptor = encrypt(key, nonce, fileSelected, (err, encryptedBlob) => {
      if (err) {
        encryptionErrors.innerText = err.message || err
        return
      }

      nonceField.value = nacl.util.encodeBase64(nonce)
      keyUsed.value = nacl.util.encodeBase64(key)
      encrypted = encryptedBlob
      downloadEncrypted.removeAttribute('disabled')
      encryptedObjURL = URL.createObjectURL(encrypted)
      downloadEncrypted.href = encryptedObjURL
    })

    encryptor.on('progress', ({ position, length }) => {
      encryptionErrors.innerText = `${(position / length) * 100}%`
    })
  }
})


example.js
/* eslint-env browser */
const naclBlob = require('./index')
const nacl = require('tweetnacl')

const key = nacl.box.keyPair().secretKey
const nonce = nacl.randomBytes(16)

const arr = new Uint8Array(10 * 1024 * 1024 + 111)
for (let i = 0; i < arr.length; i++) arr[i] = i & 0xff
const blob = new Blob([arr])

const encryptor = naclBlob.encrypt(key, nonce, blob, (err, encryptedBlob) => {
  if (err) throw (err)
  // some time later
  const decryptor = naclBlob.decrypt(key, nonce, encryptedBlob, (err, decryptedBlob) => {
    if (err) throw (err)
    compareBlobs(blob, decryptedBlob)
  })
  decryptor.on('progress', ({ position, length }) => { console.log('decrypting ' + (position / length) * 100 + '%') })
})

encryptor.on('progress', ({ position, length }) => { console.log('encrypting ' + (position / length) * 100 + '%') })

function compareBlobs (a, b) {
  const r1 = new FileReader()
  r1.onload = function () {
    const r2 = new FileReader()
    r2.onload = function () {
      if (r1.result !== r2.result) {
        console.error('blobs differ')
      } else {
        console.log('blobs are equal')
      }
    }
    r2.readAsBinaryString(b)
  }
  r1.readAsBinaryString(a)
}
